### PR TITLE
Support sending messages to remote conversations

### PR DESCRIPTION
**Pull request by pcapriotti**
This adds support for sending messages to remote conversations.

## Summary

 - Added `send-message` RPC. This takes a conversation ID, a sender, and a base64-encoded protobuf message.
 - The RPC is used when a message is sent to a remote conversation.
 - Added a federation test. This was particularly complicated to setup, because it tests federated calls which themselves make federated requests, hence require a mock federator.
 - Added an end2end test. This is pretty much identical to the other remote message sending test, but now the message is sent from the other domain, not the one that owns the conversation.
 - Enabled messaging test. This was just checking that a "federation-not-implemented" error was thrown, but now the test is enabled, and there is also an extra test which checks correct handling of failure of the federated call.
 - Re-organised sender client checks in `Galley.API.Message` so that they now work with both a local and remote sender.

## Checklist

 - [x] Title of this PR explains the impact of the change.
 - [x] The description provides context as to why the change should occur and
       what the code contributes to that effect. This could also be a link to a
       JIRA ticket or a Github issue, if there is one.
 - [x] The CHANGELOG.md file in the *Unreleased* section has been updated to
       explain the change which will be included in the release notes.
 - [x] If a component uses a new or changed internal endpoint of another
       component, this is mentioned in the CHANGELOG.md.
 - [x] If this PR creates a new endpoint, or adds a new configuration flag, the endpoint / config-flag checklist (see Wire-employee only backend [wiki page](https://github.com/zinfra/backend-wiki/wiki/Checklists)) has been followed.

### CODE REVIEW

**Review by smatting — CHANGES_REQUESTED** [review]

File: libs/wire-api-federation/src/Wire/API/Federation/Util/Aeson.hs [hunk 1]
```diff
@@ -43,3 +51,64 @@ instance (Generic a, GToJSON Zero (Rep a)) => ToJSON (CustomEncoded a) where
...
+getStatus :: forall a. HasStatus a => Int
+getStatus = statusCode (statusOf (Proxy @a))
+
+getStatuses :: forall as. All HasStatus as => [Int]
+getStatuses = unK (cpara_SList (Proxy @HasStatus) (K []) g :: K [Int] as)
+  where
+    g :: forall y ys. (HasStatus y, All HasStatus ys) => K [Int] ys -> K [Int] (y ': ys)
+    g = K . (getStatus @y :) . unK
+
+instance All UnionMember as => ToJSON (CustomEncodedUnion as) where
```
[the remark this answers is not in the document]
Comment (smatting) [reply libs/wire-api-federation/src/Wire/API/Federation/Util/Aeson.hs]: This has it's own PR now https://github.com/wireapp/wire-server/pull/1639

File: services/galley/src/Galley/API/Message.hs [hunk 2]
```diff
@@ -209,40 +266,47 @@ postQualifiedOtrMessage senderType sender mconn convId msg = runUnionT $ do
...
+        senderClient
+        (Map.findWithDefault mempty (senderDomain, senderUser) qualifiedClients)
+    )
+    $ throwError FederatedGalley.MessageNotSentUnknownClient
 
   let (sendMessage, validMessages, mismatch) =
         checkMessageClients
-          (senderDomain, sender, qualifiedNewOtrSender msg)
-          (qualifiedLocalClients <> qualifiedRemoteClients)
+          (senderDomain, senderUser, qualifiedNewOtrSender msg)
```
Comment (akshaymankar) [inline services/galley/src/Galley/API/Message.hs:283]:
```suggestion
          (senderDomain, senderUser, senderClient)
```

**Review by akshaymankar — APPROVED** [review]
Looks good now, will wait for others to review.

**Review by smatting — APPROVED** [review]

**Review by jschaul — APPROVED** [review]
Except that CI doesn't seem happy yet, the code here looks good!

**Comment by akshaymankar** [comment]
> Except that CI doesn't seem happy yet, the code here looks good!

Rebase broke it, fixed in d1149c857bdb3dc2b039f70e45d074b68b2a0f67

**Comment by pcapriotti** [comment]
Looks good to me. I guess we can actually remove `UnionT` altogether, since it's dead code now.

**Comment by akshaymankar** [comment]
> Looks good to me. I guess we can actually remove `UnionT` altogether, since it's dead code now.

Created https://github.com/wireapp/wire-server/pull/1647

**Pull request merged**